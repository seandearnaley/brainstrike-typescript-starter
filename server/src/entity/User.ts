import { Entity, PrimaryGeneratedColumn, Column } from "typeorm";

@Entity()
export class User {
  @PrimaryGeneratedColumn("uuid")
  id!: string;

  @Column()
  name!: string;

  @Column()
  email!: string;

  @Column()
  username!: string;

  @Column()
  password!: string;

  @Column({
    type: "timestamp",
    nullable: false,
    default: () => "LOCALTIMESTAMP",
  })
  created!: Date;

  @Column({
    type: "timestamp",
    nullable: false,
    default: () => "LOCALTIMESTAMP",
  })
  updated!: Date;

  // @BeforeInsert()
  // async hashPasswordBeforeInsert() {
  //   this.password = await argon2.hash(this.password, { hashLength: 12 });
  // }
}
