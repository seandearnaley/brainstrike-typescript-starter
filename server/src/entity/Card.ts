import { Entity, PrimaryGeneratedColumn, Column, ManyToMany } from "typeorm";
import { Category } from "./Category";

@Entity()
export class Card {
  @PrimaryGeneratedColumn("uuid")
  id!: string;

  @Column()
  number!: number;

  @Column()
  label!: string;

  @Column("text")
  description!: string;

  @Column({
    type: "timestamp",
    nullable: false,
    default: () => "LOCALTIMESTAMP",
  })
  created!: Date;

  @Column({
    type: "timestamp",
    nullable: false,
    default: () => "LOCALTIMESTAMP",
  })
  updated!: Date;

  @ManyToMany(() => Category, (category) => category.cards)
  categories!: Category[];
}
