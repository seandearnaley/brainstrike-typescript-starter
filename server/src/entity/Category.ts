import {
  Entity,
  Tree,
  Column,
  PrimaryGeneratedColumn,
  TreeChildren,
  TreeParent,
  ManyToMany,
  JoinTable,
} from "typeorm";

import { Card } from "./Card";

@Entity()
@Tree("closure-table")
export class Category {
  @PrimaryGeneratedColumn("uuid")
  id!: string;

  @Column()
  name!: string;

  @TreeChildren()
  children?: Category[];

  @TreeParent()
  parent?: Category;

  @Column({
    type: "timestamp",
    nullable: false,
    default: () => "LOCALTIMESTAMP",
  })
  created!: Date;

  @Column({
    type: "timestamp",
    nullable: false,
    default: () => "LOCALTIMESTAMP",
  })
  updated!: Date;

  @ManyToMany(() => Card, (card) => card.categories)
  @JoinTable()
  cards!: Card[];
}
