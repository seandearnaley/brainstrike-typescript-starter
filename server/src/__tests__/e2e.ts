import {
  GraphQLRequest,
  Observable,
  FetchResult,
  toPromise,
} from "apollo-link";

import {
  startTestServer,
  constructTestServer,
  createTestingConnection,
  DataSource,
} from "./__utils";

import * as TDATA from "./__testData";
import * as GQL from "./__queries";
import { Card, Category } from "../entity";

describe("Server - e2e", () => {
  let connection: DataSource;

  let stop: () => void,
    graphql: ({}: GraphQLRequest) => Observable<FetchResult>;

  beforeAll(async () => {
    console.log("creating test connection");
    connection = await createTestingConnection();

    // Log entity metadata status
    console.log("Card metadata exists:", connection.hasMetadata(Card));
    console.log("Category metadata exists:", connection.hasMetadata(Category));
  });

  afterAll(() => connection.close());

  beforeEach(async () => {
    const { apolloServer } = await constructTestServer(connection);
    const testServer = await startTestServer(apolloServer);
    stop = testServer.stop;
    graphql = testServer.graphql;
  });

  afterEach(async () => {
    stop();
  });

  it("gets list of cards", async () => {
    const res = await toPromise(
      graphql({
        query: GQL.GET_CARD_DATA,
        variables: { first: 20 },
      })
    );

    expect(res).toMatchSnapshot();
  });

  it("gets a single card", async () => {
    const res = await toPromise(
      graphql({
        query: GQL.GET_CARD,
        variables: { id: TDATA.mockE2EFirstId },
      })
    );
    expect(res).toMatchSnapshot();
  });
});
