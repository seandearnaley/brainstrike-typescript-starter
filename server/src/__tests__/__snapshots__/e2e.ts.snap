// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server - e2e gets a single card 1`] = `
{
  "data": {
    "card": {
      "__typename": "Card",
      "created": "2020-02-01T20:19:27.909Z",
      "description": "Amet autem et.",
      "id": "NzVmNzc0NzMtYjc0Yy00YTIwLWFlMGQtNzNlMDRiNDI3Y2RlOkNhcmQ=",
      "label": "sed possimus tempore",
      "number": 99999,
      "updated": "2020-02-01T20:19:27.909Z",
    },
  },
}
`;

exports[`Server - e2e gets list of cards 1`] = `
{
  "data": {
    "cards": {
      "edges": [
        {
          "cursor": "Q3xDYXJkfGU5MTU3NTE3LTRkNzgtNDNkZC1hMjUxLTBkNDI2OGJkNDY1ZHwx",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:05:43.666Z",
            "description": "Ipsum recusandae sit ut ut non officiis magnam aliquam nesciunt.",
            "id": "ZTkxNTc1MTctNGQ3OC00M2RkLWEyNTEtMGQ0MjY4YmQ0NjVkOkNhcmQ=",
            "label": "voluptas dolores distinctio",
            "number": 12210,
            "updated": "2020-02-02T19:05:43.666Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfGNjMTg4YzZhLTU1YTUtNDY0ZS04M2U2LWVlY2E3OGQzNGQ3NHwy",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:05:38.390Z",
            "description": "Corporis tempore eligendi est natus facilis dicta molestiae ipsa.",
            "id": "Y2MxODhjNmEtNTVhNS00NjRlLTgzZTYtZWVjYTc4ZDM0ZDc0OkNhcmQ=",
            "label": "aut exercitationem laborum",
            "number": 98246,
            "updated": "2020-02-02T19:05:38.390Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfDBhYmM1ODIxLTY5ZGMtNDE5Mi04MjRjLTVhM2M5MTRkZmFhZXwz",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:05:33.031Z",
            "description": "Aut adipisci est ipsa rerum sit dignissimos ullam.",
            "id": "MGFiYzU4MjEtNjlkYy00MTkyLTgyNGMtNWEzYzkxNGRmYWFlOkNhcmQ=",
            "label": "rerum minus iure",
            "number": 98249,
            "updated": "2020-02-02T19:05:33.031Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfGRlM2I5OGViLTQ5ODktNGRmYS04MjNjLTA3MDY5ZWZiYmE3Znw0",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:05:18.265Z",
            "description": "Molestiae occaecati neque repellendus veniam.",
            "id": "ZGUzYjk4ZWItNDk4OS00ZGZhLTgyM2MtMDcwNjllZmJiYTdmOkNhcmQ=",
            "label": "adipisci quia explicabo",
            "number": 5095,
            "updated": "2020-02-02T19:05:18.265Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfGE3ZGIxY2VhLWU3ZGMtNDVkYy05NzRiLTM0OTE5MjEyY2VkMXw1",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:04:38.922Z",
            "description": "Aut maxime quia doloribus eligendi sit enim impedit qui ab.",
            "id": "YTdkYjFjZWEtZTdkYy00NWRjLTk3NGItMzQ5MTkyMTJjZWQxOkNhcmQ=",
            "label": "iste aperiam et",
            "number": 39346,
            "updated": "2020-02-02T19:04:38.922Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfGRhMGU5MzA2LTk1ODctNGNmNC05YzRjLWRhZjE3MWFlOWJiNnw2",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:04:12.131Z",
            "description": "Sit est quidem officia.",
            "id": "ZGEwZTkzMDYtOTU4Ny00Y2Y0LTljNGMtZGFmMTcxYWU5YmI2OkNhcmQ=",
            "label": "quas molestiae esse",
            "number": 29301,
            "updated": "2020-02-02T19:04:12.131Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfDA1NTM3Zjc1LWRjMmUtNDc4Yy1iNzJlLWYyNmM2NzhlYTM5N3w3",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:04:05.010Z",
            "description": "Eos velit at est rerum ut.",
            "id": "MDU1MzdmNzUtZGMyZS00NzhjLWI3MmUtZjI2YzY3OGVhMzk3OkNhcmQ=",
            "label": "est voluptate et",
            "number": 34517,
            "updated": "2020-02-02T19:04:05.010Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfGZhOGZmOTM2LTI5MzItNDU3ZS04ZjMzLTg4YjY5Mzg0YTc1MHw4",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:03:57.878Z",
            "description": "Commodi eius voluptas dolores iste ratione voluptas quis.",
            "id": "ZmE4ZmY5MzYtMjkzMi00NTdlLThmMzMtODhiNjkzODRhNzUwOkNhcmQ=",
            "label": "repellat et quia",
            "number": 42478,
            "updated": "2020-02-02T19:03:57.878Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfDE2NmZjNmE1LWEwMTQtNDk1MC1hNGY3LWUyMWQ4ZjlmMWEyMnw5",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:03:54.906Z",
            "description": "Voluptas vitae corporis ab quas tempora.",
            "id": "MTY2ZmM2YTUtYTAxNC00OTUwLWE0ZjctZTIxZDhmOWYxYTIyOkNhcmQ=",
            "label": "officiis quibusdam ipsam",
            "number": 8553,
            "updated": "2020-02-02T19:03:54.906Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfDJjZTRhOGExLTdlNjAtNGYyZS05NDBiLWMzZTM3Y2M4ZjQzMnwxMA==",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:03:53.654Z",
            "description": "Unde incidunt quia maxime et.",
            "id": "MmNlNGE4YTEtN2U2MC00ZjJlLTk0MGItYzNlMzdjYzhmNDMyOkNhcmQ=",
            "label": "omnis dolorem iusto",
            "number": 80662,
            "updated": "2020-02-02T19:03:53.654Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfDE3MjBjZjc0LTdkOTItNGJmYS05MTBiLTc1MGMyZWNmYzdlN3wxMQ==",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:03:38.174Z",
            "description": "Rerum consequatur dolorem qui eum sit.",
            "id": "MTcyMGNmNzQtN2Q5Mi00YmZhLTkxMGItNzUwYzJlY2ZjN2U3OkNhcmQ=",
            "label": "facilis necessitatibus atque",
            "number": 92561,
            "updated": "2020-02-02T19:03:38.174Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfDcyYTRiN2MzLWYxNjEtNDg1OS1iMzI4LTUzYjM3OTE5MzliMHwxMg==",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:03:33.566Z",
            "description": "Rerum distinctio fuga asperiores et.",
            "id": "NzJhNGI3YzMtZjE2MS00ODU5LWIzMjgtNTNiMzc5MTkzOWIwOkNhcmQ=",
            "label": "eos sint placeat",
            "number": 33375,
            "updated": "2020-02-02T19:03:33.566Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfGIzOTE3MTk5LTAyODEtNGYxMC1iMWJlLTFjYmMwZmIyYWMxN3wxMw==",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:03:24.499Z",
            "description": "Voluptatem quo sunt.",
            "id": "YjM5MTcxOTktMDI4MS00ZjEwLWIxYmUtMWNiYzBmYjJhYzE3OkNhcmQ=",
            "label": "deleniti et qui",
            "number": 45499,
            "updated": "2020-02-02T19:03:24.499Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfGM5ZGE2MzRmLTk4NjktNGI0MS1hY2FiLThhM2IzZTRmZjQxMnwxNA==",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:03:22.659Z",
            "description": "Ab nostrum incidunt.",
            "id": "YzlkYTYzNGYtOTg2OS00YjQxLWFjYWItOGEzYjNlNGZmNDEyOkNhcmQ=",
            "label": "culpa vitae est",
            "number": 38482,
            "updated": "2020-02-02T19:03:22.659Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfGEzYzdhYTU1LWM4ZWYtNDczMS05NDNlLTlhNTdjMjA2NTcxYXwxNQ==",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:03:16.177Z",
            "description": "Deleniti voluptatem nesciunt atque dolores.",
            "id": "YTNjN2FhNTUtYzhlZi00NzMxLTk0M2UtOWE1N2MyMDY1NzFhOkNhcmQ=",
            "label": "quae ullam illo",
            "number": 88443,
            "updated": "2020-02-02T19:03:16.177Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfDVkMWUyOTIzLTM3MDMtNDcxZi1iZTNlLWY1ZWJhNmZiNTlmY3wxNg==",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:03:11.265Z",
            "description": "Dolorum at quia quasi id rem consectetur non dolorum.",
            "id": "NWQxZTI5MjMtMzcwMy00NzFmLWJlM2UtZjVlYmE2ZmI1OWZjOkNhcmQ=",
            "label": "cumque ipsam ut",
            "number": 14665,
            "updated": "2020-02-02T19:03:11.265Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfGJjYzAzMjg2LTkyM2EtNDNkNi1hYTBkLTMyODVkNTJjZDk0ZHwxNw==",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:02:59.693Z",
            "description": "Qui rerum nemo suscipit rerum voluptatem.",
            "id": "YmNjMDMyODYtOTIzYS00M2Q2LWFhMGQtMzI4NWQ1MmNkOTRkOkNhcmQ=",
            "label": "ducimus laboriosam porro",
            "number": 68703,
            "updated": "2020-02-02T19:02:59.693Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfDJmMjRmNjg4LTEzMmMtNDFjZC04NjkwLWY5MjNlMWJkYTc0MXwxOA==",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:02:51.052Z",
            "description": "Non debitis eos inventore dolorum quae dolor est aspernatur praesentium.",
            "id": "MmYyNGY2ODgtMTMyYy00MWNkLTg2OTAtZjkyM2UxYmRhNzQxOkNhcmQ=",
            "label": "unde odit laudantium",
            "number": 88574,
            "updated": "2020-02-02T19:02:51.052Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfGQxNDFiMTRlLTFmOGItNDQwZS05YjUxLTllMTBlN2ZkYmMzZHwxOQ==",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:02:46.467Z",
            "description": "Veritatis deserunt corrupti maxime est.",
            "id": "ZDE0MWIxNGUtMWY4Yi00NDBlLTliNTEtOWUxMGU3ZmRiYzNkOkNhcmQ=",
            "label": "sed modi maxime",
            "number": 67971,
            "updated": "2020-02-02T19:02:46.467Z",
          },
        },
        {
          "cursor": "Q3xDYXJkfDQxYjZkM2U4LTk4NjgtNDMyMC04ZDNlLWJlZTAzMDU1Y2ZmOHwyMA==",
          "node": {
            "__typename": "Card",
            "created": "2020-02-02T19:02:22.508Z",
            "description": "Expedita est rem maxime sunt unde adipisci.",
            "id": "NDFiNmQzZTgtOTg2OC00MzIwLThkM2UtYmVlMDMwNTVjZmY4OkNhcmQ=",
            "label": "labore sapiente fugiat",
            "number": 36728,
            "updated": "2020-02-02T19:02:22.508Z",
          },
        },
      ],
      "pageInfo": {
        "endCursor": "Q3xDYXJkfDQxYjZkM2U4LTk4NjgtNDMyMC04ZDNlLWJlZTAzMDU1Y2ZmOHwyMA==",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "Q3xDYXJkfGU5MTU3NTE3LTRkNzgtNDNkZC1hMjUxLTBkNDI2OGJkNDY1ZHwx",
        "totalCount": 10000,
      },
    },
  },
}
`;
